### ISSUE TITLE
Finish implementing occlusion buffer in software renderer

### ISSUE THREAD

**Issue by afritz1** [afritz1/OpenTESArena#120]
I tried implementing an occlusion buffer several months ago but I couldn't get it to work, so if someone would like to finish it, I would appreciate it.

Basically what it does is allow the renderer to draw front to back with no voxel overdraw and removes the need for checking per-pixel depth with opaque walls. Since there's a ton of overdraw at the moment, it should be a noticeable performance boost once it's done.

Each entry in the `SoftwareRenderer::occlusion` vector is like a window for a column in the frame buffer; `yMin` is the inclusive top of the window and `yMax` is the exclusive bottom.
https://github.com/afritz1/OpenTESArena/blob/17790212531ec17def25f31ee1522e3af5e98bc5/OpenTESArena/src/Rendering/SoftwareRenderer.h#L81
The window gets smaller when a drawing range (a.k.a., a new column of pixels) covers one or both of those values. When the window is completely shut then no more voxels are rendered in that column and the ray caster can return early.

The method calls are already in the right place; it's just the methods themselves that need implementing:
https://github.com/afritz1/OpenTESArena/blob/17790212531ec17def25f31ee1522e3af5e98bc5/OpenTESArena/src/Rendering/SoftwareRenderer.cpp#L139
https://github.com/afritz1/OpenTESArena/blob/17790212531ec17def25f31ee1522e3af5e98bc5/OpenTESArena/src/Rendering/SoftwareRenderer.cpp#L160

If you'd like more details about occlusion buffers (that's what I'm calling it anyway; not sure that it's _actually_ occlusion culling), then look at [this page](http://fabiensanglard.net/duke3d/build_engine_internals.php) on how it's done in Duke Nukem 3D (search for the `Wall/Ceiling/Floor rendition` section). It's basically the same thing as what I'm trying to do.

**Comment by enderdrag64** [comment]
Replacing 
```C++
	else if (canIncreaseMin)
	{
		this->yMin = std::min(yEnd, this->yMax);
	}
	else if (canDecreaseMax)
	{
		this->yMax = std::max(yStart, this->yMin);
	} 
```
with 
```C++
	else if (canIncreaseMin)
	{
		this->yMin = std::max(yStart, this->yMin);
	}
	else if (canDecreaseMax)
	{
		this->yMax = std::min(yEnd, this->yMax);
	}
```
should fix the occlusion buffer.

**Comment by afritz1** [comment]
Thanks for looking into this (continuing from our Discord conversation).

It turns out that your changes work (in a way) because they turn the incorrect assignments into no-ops. Unfortunately, the two else ifs there are now always assigning to themselves.

So the primary occlusion case is working (when a pixel column covers an entire screen column; i.e., `yStart <= this->yMin && yEnd >= this->yMax`), but the other two cases are not (when it's just a portion of the screen being occluded, i.e., from a floor or ceiling).

Basically, commenting out those two else ifs is the same as using your changes. I'd like to spend more time on this soon if I can so it can get wrapped up.

See commit b36271eee0d8687b80c0548514134a46e4c78ba6.

**Comment by enderdrag64** [comment]
You're right I forgot to swap the start and end values. Now these cases should work:
```C++
	else if (canIncreaseMin)
	{
		this->yMin = std::max(yEnd, this->yMin);
	}
	else if (canDecreaseMax)
	{
		this->yMax = std::min(yStart, this->yMax);
	}
```

**Comment by afritz1** [comment]
Looks like it's all working now, thanks a lot!
